### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by hyc**
Conclusion of perf work started by warptangent. Significant space and runtime reductions, as noted here https://gist.github.com/hyc/33f3eec6bae83246209d

### CODE REVIEW

File: src/blockchain_db/lmdb/db_lmdb.h [hunk 1]
```diff
@@ -39,15 +39,20 @@
 namespace cryptonote
 {
 
+typedef struct mdb_block_info
+{
+  uint64_t bi_timestamp;
+  uint64_t bi_coins;
+  size_t bi_size;
```
Comment (moneromooo-monero) [inline src/blockchain_db/lmdb/db_lmdb.h]: Do we want to switch to uint64_t here ?

File: src/blockchain_db/lmdb/db_lmdb.h [hunk 2]
```diff
@@ -39,15 +39,20 @@
...
+  difficulty_type bi_diff;
```
Comment (moneromooo-monero) [inline src/blockchain_db/lmdb/db_lmdb.h]: and possibly same here

File: src/blockchain_db/lmdb/db_lmdb.h (hunk 1 above)
Comment (hyc) [inline src/blockchain_db/lmdb/db_lmdb.h]: Yes, this broke 32-bit compatibility. Fixed already in following commit https://github.com/monero-project/bitmonero/pull/794/commits/bef1bbf5ea8ea6055046363fcb326e9fa4e68393

File: src/blockchain_db/lmdb/db_lmdb.h (hunk 2 above)
Comment (hyc) [inline src/blockchain_db/lmdb/db_lmdb.h]: difficulty_type -- difficulty.h:    typedef std::uint64_t difficulty_type;
Shouldn't that already be the same as uint64_t?

Comment (hyc) [inline src/blockchain_db/lmdb/db_lmdb.h]: difficulty_type -- difficulty.h:    typedef std::uint64_t difficulty_type;
Shouldn't that already be the same as uint64_t?

**Comment by iamsmooth** [comment]
Do we want to merge this straight into master without there being a major release imminent (or is there one?). It seems pretty disruptive to roll out a DB format change (requiring rsync or convert?) in between major releases when often small bug fixes are needed for point releases or by people who pull mid-release.

**Comment by hyc** [comment]
Agreed, it is disruptive since it requires a full resync. But I also believe the benefit is significant, and I don't want it to bitrot in a side branch.

**Comment by othexmr** [comment]
I am for merging this and bumping the version to 0.9.2 after other minor issues and bugfixes are applied.

Differences on my tests are so significant i would not use the masterbranch anymore without this beeing merged.

**Comment by iamsmooth** [comment]
"I am for merging this and bumping the version to 0.9.2 after other minor issues and bugfixes are applied."

I agree with this but I think it should happen in the other order. When the other minor issues are resolved and bug fixes are applied, then we should merge this and release 0.9.2.

**Comment by iamsmooth** [comment]
Or some release number (0.9.2 is wrong).

**Comment by Gingeropolous** [comment]
I would argue its less disruptive now when everyone is fully aware of all of these recent point releases, and when the network is relatively small, and services are aware that they should keep their eye on Monero. 

Also, re: full resync / conversion - I still think packaging the update with a conversion script would be adequate. No fancy conversion tool necessary. Include binaries that export the old one to raw, and then import to the new format. Can assume verify off because blockchain is local. One command and the user can update to new version without massive bandwidth costs.

**Comment by iamsmooth** [comment]
Upgrade script that exports and imports with no verify sounds fine to me.

**Comment by hyc** [comment]
ok. for the nitty gritty details of an upgrade script - would we just include an old blockchain_export-0.9.4 with the new binaries? Or ask the user for the location of their old binaries?

**Comment by fluffypony** [comment]
Going back to the discussion a few dev meetings ago, it would be nice to have a migrations-style setup. Basically each migration is a function that knows how to convert (via import/export or whatever) from version N-1 to version N. So if I was on blockchainDB version 1, and now I download Monero and it uses blockchainDB version 5, then it just runs the functions, sequentially, to go from v1 -> v2 -> v3 -> v4 -> v5.

The migration can be self-contained, in that it can just use the previous code as-is without needing to maintain it (since old version formats remain fixed).

**Comment by Gingeropolous** [comment]
> ok. for the nitty gritty details of an upgrade script - would we just include an old blockchain_export-0.9.4 with the new binaries? Or ask the user for the location of their old binaries?

I would include new binaries. Can't assume the user either 1) knows where they are or 2) didn't delete them for whatever reason. Easiest solution is to include a wget / download (does DOS/windows have a wget?) of the proper binary for the needed conversion. Flaws - assumes binary compatibility with OS. Introduces dependency of centralized server service, but script could have fallback "conversion tool not available - proceeding to resync from network?". 

> So if I was on blockchainDB version 1, and now I download Monero and it uses blockchainDB version 5, then it just runs the functions, sequentially, to go from v1 -> v2 -> v3 -> v4 -> v5.

Indeed, so it will pull the proper bins from the server and then execute. Though if the raw format remains forward compatible, then may not need intermediates? I dunno.

**Comment by iamsmooth** [comment]
Everyone agrees that migration would be nice but the question right now is whether to hold the improved DB implementation until migration gets implemented and tested. The rough consensus seems to be that the improvement is huge enough to roll out ASAP.

**Comment by moneromooo-monero** [comment]
If it's not too much hassle for fluffypony, I'd merge all fixes in a release, then just this DB change in a second one, so people who may not want to update DB right away can still get all else.

About downloading binaries with wget, no. Either add them to a release (maybe renamed to "convert-dnb-to-0.9.50-format, even), or ask the user to do it manually, because signatures and I hate stuff that wants to download and run random things.

**Comment by hyc** [comment]
Not sure if we should continue the migration discussion here or on #663 but... if we continue down that route instead of just providing a script that does blockchain_export/blockchain_import, should migration start automatically when the daemon starts, or should it be a manually invoked command? Since it may be quite time consuming, I'd suggest manually invoked. Perhaps we can decide that later, after we have a migration function working.

**Comment by fluffypony** [comment]
@hyc it's been my experience that:

- new versions of the daemon need the latest version of the database
- lots of daemons (most of them, really) are virtually unattended
- there are a significant number of Windows users using the baked-in Windows service functionality
- thus the majority of people won't realise that their newly updated daemon is dying on startup

The daemon needs to be as robust as is humanly possible - nothing should stop it from starting, recovering, and running, least of all a migration that may only take a minute (in some circumstances).

**Comment by antanst** [comment]
In my humble opinion, requiring a _manual_ DB conversion in between "stable" releases is a sure way to surprise users and make them wary of future upgrades.

**Comment by hyc** [comment]
IMO, doing it silently and automagically is a mistake. If users are unaware that the DB format has changed, and then for whatever reason decide to rollback to an earlier software version, their DB will be unusable and they won't know why. I'd say this requires explicit knowledge - release notes and various warnings. This is also another point in favor of just exporting to blockchain.raw and reimporting - then they still have their old DB as a backup, in usable form. Currently the migration code I'm working on deletes the old format records immediately after inserting the new format into the DB, so once a migration starts it's a one-way trip. It can be interrupted and resumed but the old code won't be able to use it any more.

**Comment by iamsmooth** [comment]
I agree with both of the previous points and do not think the DB format should change (in an incompatible way) on minor releases. If we are pushing this out it should be some sort of major-ish release.

**Comment by luigi1111** [comment]
I agree with the three previous posts.

Getting to the conversion stuff, are we going to be changing format enough to warrant significant effort towards a slick conversion path that covers X number of edge cases, etc?
Maybe we could have a simple check for any old database and change the folder name to *.V# and just restart syncing from scratch (assuming enough free space), of course along with obvious release notes.

Back to this PR, I think we could maybe do some beta version of it to get more users using it and leave time for any new incompatible ideas that may come up (though the latter is really up to hyc's feeling on the matter).

**Comment by iamsmooth** [comment]
"Maybe we could have a simple check for any old database and change the folder name to *.V# and just restart syncing from scratch (assuming enough free space), of course along with obvious release notes."

Pretty reasonable suggestion that also means that future format upgrades become much easier, since they don't require custom conversion code every time. Also addresses the issue of downgrading.

**Comment by hyc** [comment]
I'm fine with leaving it in a beta version for a while longer. Also I've added an auto-migrate function to this branch. You can try it out and see if it's tolerable.

**Comment by fluffypony** [comment]
We can always trigger an automigrate under the two obvious circumstances: we're in service mode on Windows, or we're in background mode on other operating systems?

**Comment by hyc** [comment]
I have an alternate migrate implementation to test in https://github.com/LMDB/bitmonero/tree/migr4b

It deletes a bunch of the old indices and recreates them instead of trying to migrate them in-place. Was faster in my own testing.

**Pull request merged**